dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/relationship/JacksonRelationshipService.java

```diff
@@ -0,0 +1,277 @@
+    {
+        if ( importOptions == null )
+        {
+            importOptions = new ImportOptions();
+        }
+
+        if ( importOptions.getUser() == null )
+        {
+            importOptions.setUser( currentUserService.getCurrentUser() );
+        }
+
+        return importOptions;
+    }
+
+    private ImportSummaries addRelationshipList( List<Relationship> relationships, ImportOptions importOptions )
+    {
+        ImportSummaries importSummaries = new ImportSummaries();
+        importOptions = updateImportOptions( importOptions );
+
+        List<Relationship> create = new ArrayList<>();
+        List<Relationship> update = new ArrayList<>();
+        List<Relationship> delete = new ArrayList<>();
+
+        if ( importOptions.getImportStrategy().isCreate() )
+        {
+            create.addAll( relationships );
+        }
+        else if ( importOptions.getImportStrategy().isCreateAndUpdate() )
+        {
+            for ( Relationship relationship : relationships )
+            {
+                sortCreatesAndUpdates( relationship, create, update );
+            }
+        }
+        else if ( importOptions.getImportStrategy().isUpdate() )
+        {
+            update.addAll( relationships );
+        }
+        else if ( importOptions.getImportStrategy().isDelete() )
+        {
+            delete.addAll( relationships );
+        }
+        else if ( importOptions.getImportStrategy().isSync() )
+        {
+            for ( Relationship relationship : relationships )
+            {
+                sortCreatesAndUpdates( relationship, create, update );
+            }
+        }
+
+        importSummaries.addImportSummaries( addRelationships( create, importOptions ) );
+        importSummaries.addImportSummaries( updateRelationships( update, importOptions ) );
+        importSummaries.addImportSummaries( deleteRelationships( delete, importOptions ) );
+
+        if ( ImportReportMode.ERRORS == importOptions.getReportMode() )
+        {
+            importSummaries.getImportSummaries().removeIf( is -> is.getConflicts().isEmpty() );
+        }
+
+        return importSummaries;
+    }
+
+    private void sortCreatesAndUpdates( Relationship relationship, List<Relationship> create,
+        List<Relationship> update )
+    {
+        if ( StringUtils.isEmpty( relationship.getRelationship() ) )
+        {
+            create.add( relationship );
+        }
+        else
+        {
+            if ( !relationshipService.relationshipExists( relationship.getRelationship() ) )
+            {
+                create.add( relationship );
+            }
+            else
+            {
+                update.add( relationship );
+            }
+        }
+    }
+
+    @SuppressWarnings( "unchecked" )
+    private static <T> T fromXml( InputStream inputStream, Class<?> clazz )
+        throws IOException
+    {
+        return (T) XML_MAPPER.readValue( inputStream, clazz );
```

Comment:
Generic methods - can we re-use from somewhere else?

Reply:
Same as previous comment :)